scheduler/src/ratelimit.clj

```diff
@@ -0,0 +1,114 @@
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns ratelimit
+  (:require [ratelimit.token-bucket-filter]
+            [clj-time.core]
+            [clj-time.coerce])
+  (:import (com.google.common.cache CacheBuilder Cache)
+           (java.util.concurrent TimeUnit)))
+
+(defprotocol RateLimiter
+  (earn-tokens! [this key]
+    "Account for any earned tokens for the requested key.")
+  (in-debt!? [this key]
+    "Boolean indicating if the rate limit is currently in debt. Earns tokens first, then checks to see if in debt.")
+  (spend-tokens! [this key tokens]
+    "Request this number of tokens. Does not block (or otherwise indicate we're in debt")
+  (time-until!? [this key]
+    "Time, in milliseconds, until this rate limiter is out of debt for this key. Earns tokens first.")
+  (is-enforcing? [this]
+    "Are we enforcing this policy?"))
+
+(defn now
+  []
+  (clj-time.coerce/to-long (clj-time.core/now)))
+
+(defn get-tbf
+  "Get or create a TBF for the requested key. Buckets start off full, so users whose first request is a lot of
+  fast requests don't get throttled immediately. This doesn't let users get much above their quota ---
+  The fact that we don't have a TBF says that we are either just starting up (not likely) or haven't
+  gotten anything from them in a long time. (Assuming the TTL is set at least as high as
+  (:bucket-size/:rate-per-minute)."
+  [limiter key]
+  (.get (:cache limiter) key (proxy [Callable] []
+                               (call []
+                                 (ratelimit.token-bucket-filter/create-tbf (/ (:rate-per-minute limiter) 60000.)
+                                                                     (:max-tokens limiter)
+                                                                     (now)
+                                                                     (:max-tokens limiter))))))
+
+(defrecord TBF-RateLimiter
+  [^long max-tokens ^double rate-per-minute cache ^Boolean is-enforcing]
+  RateLimiter
+  (earn-tokens!
+    [this key]
+    (locking cache
+      (let [tbf (ratelimit.token-bucket-filter/earn-tokens (get-tbf this key) (now))]
+        (.put cache key tbf))))
+
+  (in-debt!?
+    [this key]
+    (earn-tokens! this key)
+    (ratelimit.token-bucket-filter/in-debt? (get-tbf this key)))
+
+  (spend-tokens!
+    [this key tokens]
+    (locking cache
+      (let [tbf (ratelimit.token-bucket-filter/spend-tokens (get-tbf this key) (now) tokens)]
+        (.put cache key tbf)
+        nil)))
+
+  (time-until!?
+    [this key]
+    (earn-tokens! this key)
+    (ratelimit.token-bucket-filter/time-until (get-tbf this key)))
+  (is-enforcing?
```

Comment:
Missing leading newline.

Reply:
Decided to be more consistent and strip out newlines everywhere here.